languages/python/sqlalchemy-oso/sqlalchemy_oso/roles.py

```diff
@@ -1,16 +1,142 @@
 from typing import Any, List
 
 from sqlalchemy.types import Integer, String
-from sqlalchemy.schema import Table, Column, ForeignKey
+from sqlalchemy.schema import Column, ForeignKey
 from sqlalchemy.ext.declarative import declared_attr
-from sqlalchemy.orm import relationship, backref
-from sqlalchemy import inspect
-
+from sqlalchemy.orm import relationship, validates
+from sqlalchemy import inspect, UniqueConstraint
+from .session import _OsoSession
 
 ROLE_CLASSES: List[Any] = []
 
 
+def resource_role_class(
+    declarative_base, user_model, resource_model, role_choices, mutually_exclusive=True
+):
+    """Create a :ref:`resource-specific role<resource-specific-roles>` Mixin
+    for SQLAlchemy models. The role mixin is an
+    `Association Object <https://docs.sqlalchemy.org/en/13/orm/basic_relationships.html#association-object>`_
+    between the ``user_model`` and the ``resource_model``.
+
+
+    :param declarative_base: The SQLAlchemy declarative base model that \
+    the role model and all related models are mapped to.
+
+    :param user_model: The SQLAlchemy model representing users that the \
```

Comment:
I know we spoke about groups being a follow up. But is there any reason this couldn't be `actor_model` and be for groups too?

Reply:
Where we wouldn't provide anything out of the box for people to get groups working with users, but at least it would be consistent with writing `allow(actor, ...)` rules for them

Reply:
Sure, we could make it `actor_model`. But then the `actor` passed into the policy would need to be the type of whatever `actor_model` they pass in here.

Reply:
Someone could write their own:

```
allow(user, action, resource) if
   group in user.groups and
    allow(group, action, resource);
```

Reply:
I'll add this as a follow up